### PR TITLE
[TEVA-2461] Changes to Applicant Counting

### PR DESCRIPTION
**Pull request by josephhull676**
## Jira ticket URL

https://dfedigital.atlassian.net/browse/TEVA-2461

## Changes in this PR:

- Removed text related to applications for scheduled jobs
- The number of applications shown now only counts submitted applications
- Added link to publisher job applications index page for jobs that have not received any applications or have only received withdrawn applications

## Comments:

- Have not done 1 and 1a (in the acceptance criteria) as this had already been done.

### No Applicants / Only Withdrawn Applicants

Screenshot 2021-04-27 at 17 06 17

### Scheduled Jobs

Screenshot 2021-04-27 at 16 13 48

### CODE REVIEW

File: app/components/publishers/vacancies_component.rb [hunk 1]
```diff
@@ -46,13 +46,15 @@ def view_applicants(vacancy, card)
     return unless vacancy.enable_job_applications?
     return unless include_job_applications?
 
-    if vacancy.job_applications.any?
-      link = govuk_link_to(I18n.t("jobs.manage.view_applicants", count: vacancy.job_applications.count),
+    if vacancy.job_applications.submitted.any?
```
Comment (cpjmcquillan) [inline app/components/publishers/vacancies_component.rb:49]: The ticket doesn't make this clear, but the count should include all job applications that don't have a status of draft or withdrawn

File: app/components/publishers/vacancies_component/vacancies_component.html.slim [hunk 2]
```diff
@@ -31,7 +31,7 @@
 
               = f.hidden_field :jobs_type, value: @selected_type
 
-      .moj-filter-layout__content class=grid_column_class
+      .moj-filter-layout__content class="grid_column_class"
```
Comment (cpjmcquillan) [inline app/components/publishers/vacancies_component/vacancies_component.html.slim:34]: Why is this change here?

File: app/components/publishers/vacancies_component.rb [hunk 3]
```diff
@@ -46,13 +46,15 @@ def view_applicants(vacancy, card)
...
+      link = govuk_link_to(I18n.t("jobs.manage.view_applicants", count: vacancy.job_applications.submitted.count),
                            organisation_job_job_applications_path(vacancy.id),
                            class: "govuk-link--no-visited-state")
       tag.div(card.labelled_item(I18n.t("jobs.manage.applications"), link))
-    elsif vacancy.job_applications.none?
-      text = tag.span(I18n.t("jobs.manage.view_applicants", count: 0))
+    elsif vacancy.job_applications.withdrawn.any? || vacancy.job_applications.none?
```
Comment (david-mears-dfe) [inline app/components/publishers/vacancies_component.rb:54]: This seems wrong - if there are 55 submitted applications, and 1 withdrawn application, this will evaluate to true, and it will say '0 applications'.

File: app/components/publishers/vacancies_component/vacancies_component.html.slim (hunk 2 above)
Comment (josephhull676) [reply app/components/publishers/vacancies_component/vacancies_component.html.slim:34]: I just happened to come across it whilst I was poking around in the file. The class name should be wrapped in double quotes, no?

File: app/components/publishers/vacancies_component.rb (hunk 3 above)
Comment (josephhull676) [reply app/components/publishers/vacancies_component.rb:54]: `if vacancy.job_applications.submitted.any?` comes first, so wouldn't it display the number of submitted applications if there are any?

File: app/components/publishers/vacancies_component/vacancies_component.html.slim (hunk 2 above)
Comment (david-mears-dfe) [reply app/components/publishers/vacancies_component/vacancies_component.html.slim:34]: This is not a class name, it's a method - I presume, otherwise it would be throwing errors.

File: app/components/publishers/vacancies_component.rb (hunk 3 above)
Comment (david-mears-dfe) [reply app/components/publishers/vacancies_component.rb:54]: I see, OK.

Comment (david-mears-dfe) [reply app/components/publishers/vacancies_component.rb:54]: What is the difference between `elsif vacancy.job_applications.withdrawn.any? || vacancy.job_applications.none?` and `else`?

File: app/components/publishers/vacancies_component.rb [hunk 4]
```diff
@@ -100,4 +102,8 @@ def set_organisation_options
       OpenStruct.new({ id: organisation.id, name: "Trust head office", label: "Trust head office (#{count})" }),
     )
   end
+
+  def post_submission_job_applications(job_applications)
```
Comment (david-mears-dfe) [inline app/components/publishers/vacancies_component.rb:106]: This could be a scope - see the scopes in vacancy.rb for examples.

**Review by david-mears-dfe — COMMENTED** [review]
What do you think, does this need any additional tests in vacancies_component_spec.rb?

File: app/components/publishers/vacancies_component.rb [hunk 5]
```diff
@@ -46,13 +46,15 @@ def view_applicants(vacancy, card)
     return unless vacancy.enable_job_applications?
     return unless include_job_applications?
 
-    if vacancy.job_applications.any?
-      link = govuk_link_to(I18n.t("jobs.manage.view_applicants", count: vacancy.job_applications.count),
+    if post_submission_job_applications(vacancy.job_applications).present?
+      link = govuk_link_to(I18n.t("jobs.manage.view_applicants", count: post_submission_job_applications(vacancy.job_applications).count),
```
Comment (david-mears-dfe) [inline app/components/publishers/vacancies_component.rb:50]: Minor performance improvement:

If you use an instance variable, you can avoid having to run the `post_submission_job_applications` database query twice.

That instance variable could be set in a controller helper_method or in the controller action itself.

File: app/components/publishers/vacancies_component.rb [hunk 6]
```diff
@@ -43,16 +43,19 @@ def no_jobs_text
   end
 
   def view_applicants(vacancy, card)
-    return unless vacancy.enable_job_applications?
-    return unless include_job_applications?
+    return unless vacancy.enable_job_applications? && include_job_applications?
```
Comment (cesidio) [inline app/components/publishers/vacancies_component.rb:46]: The reason I made this in 2 lines is because it is more readable. unless with an and or an or is more complicated to reason in your mind

Comment (josephhull676) [reply app/components/publishers/vacancies_component.rb:46]: I felt guilty about adding another line so I put this on to one line to compensate for it lol.

I don't think the one line is more readable, so I'll change it back to what it was.

Comment (cesidio) [reply app/components/publishers/vacancies_component.rb:46]: It would be more readable if it was an `if`

File: app/components/publishers/vacancies_component.rb (hunk 3 above)
Comment (josephhull676) [reply app/components/publishers/vacancies_component.rb:55]: `  enum status: { draft: 0, submitted: 1, reviewed: 2, shortlisted: 3, unsuccessful: 4, withdrawn: 5 }`

I guess nothing lol. Will replace with an else.

File: app/components/publishers/vacancies_component.rb [hunk 7]
```diff
@@ -46,13 +46,17 @@ def view_applicants(vacancy, card)
     return unless vacancy.enable_job_applications?
     return unless include_job_applications?
 
-    if vacancy.job_applications.any?
-      link = govuk_link_to(I18n.t("jobs.manage.view_applicants", count: vacancy.job_applications.count),
+    after_submission_job_applications = vacancy.job_applications.after_submission
+
+    if after_submission_job_applications.present?
```
Comment (cesidio) [inline app/components/publishers/vacancies_component.rb:51]: when it's a collection, I prefer `any?` to `present?`, it reads better

**Review by cesidio — APPROVED** [review]
